Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1333 @@
+
+A **production rule** is a rule describing how to "produce" a string that is a sentence corresponding to the grammar.
+It has form `a : b { action }`, which looks like so in Happy:
+
+```happy
+myNonTerminal
+  : bodyA1 bodyA2 ... { actionA }
+  | bodyB1 bodyB2 ... { actionB }
+  | ...
+```
+
+n.b.: The ellipses here are not correct Happy grammar, and we use them to represent that more symbols may be present.
+
+`body@#` where `@` is some letter and `#` some number means that the parser must parse each body in succession.
+The pipe (`|`) represents an alternation. The parser may parse any of the given bodies, accepting the one that matches.
+An alternative may be empty, meaning that `myNonTerminal : { action }` is also accepted.
+`action@` represents what the parser should do once successfully parsed that rule. Like in Alex, it may be any Haskell expression. More about this later.
+
+Production rules may also have an optional type signature.
+For instance, the declaration of `myNonTerminal` could also have been written as such:
+
+```happy
+myNonTerminal :: { MyType }
+  : bodyA1 bodyA2 ... { actionA }
+  | bodyB1 bodyB2 ... { actionB }
+  | ...
+```
+
+Where any Haskell type may replace `MyType`.
+
+Later in the article, we will also use the following notation representing the same as above.
+This is not valid Happy code, but it's what it uses for debugging information.
+
+```
+myNonTerminal -> bodyA1 bodyA2 ...
+myNonTerminal -> bodyB1 bodyB2 ...
+myNonTerminal -> ...
+```
+
+There may be multiple productions in a single Happy file.
+
+### Parser position
+
+It's essential to introduce a notation that will be used later in this article.
+We use a dot (`.`) to represent the current position of the parser.
+For example, suppose we have the following:
+
+```
+exp -> exp . '+' exp
+```
+
+In this case, the parser has finished consuming the first `exp`, and it's about to consume the `+`.
+
+### Look-ahead
+
+The **look-ahead** represents the sequence of symbols that will be consumed next by the parser.
+Happy is an LALR(1) parser, which means its look-ahead is 1, so it can see one incoming symbol.
+In the `exp` example above, the look-ahead token will be `+`.
+
+### Actions (shift, reduce, goto, accept)
+
+When working with Happy, two core concepts that you need to know are **shifting** and **reducing**, which I'll briefly explain.
+
+An LR parser also contains a state machine, similarly to Alex.
+This finite-state machine is called an _LR automaton_, which is also a DFA, which is used to make parsing decisions.
+At the same time, the parser contains a stack with all the tokens consumed during parsing, making it a _pushdown automaton_.
+The parser produces an action table describing how to make decisions based on its stack and the look-ahead token.
+
+**Shifting** means that the parser should push (shift) the current input symbol on the top of its internal stack, change its state, and continue parsing.
```

Comment:
How about a table or list with: "Here are the (kinds of) acceptable actions for the parser"

Reply:
This would be more structure, I think it is easier to understand them once they are all together.